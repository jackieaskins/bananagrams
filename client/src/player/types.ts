export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
