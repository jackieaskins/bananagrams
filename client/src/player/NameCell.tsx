import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
