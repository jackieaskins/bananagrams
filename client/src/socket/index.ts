import { io } from 'socket.io-client';

import { GameInfo } from '../games/types';

export type Callback<T> = (error: { message: string } | null, data: T) => void;
export type Listener<T = void> = (data: T) => void;

export const socket = io();

export const getUserId = (): string => socket.id;

export const createGame = (
  props: {
    gameName: string;
    username: string;
    isShortenedGame: boolean;
  },
  callback: Callback<GameInfo>
): void => {
  socket.emit('createGame', props, callback);
};
export const joinGame = (
  props: { gameId: string; username: string },
  callback: Callback<GameInfo>
): void => {
  socket.emit('joinGame', props, callback);
};
export const setReady = (props: { isReady: boolean }): void => {
  socket.emit('ready', props);
};
export const kickPlayer = (props: { userId: string }): void => {
  socket.emit('kickPlayer', props);
};
export const leaveGame = (props: { gameId: string }): void => {
  socket.emit('leaveGame', props);
};

export const addGameInfoListener = (listener: Listener<GameInfo>): void => {
  socket.on('gameInfo', listener);
};
export const removeGameInfoListener = (): void => {
  socket.off('gameInfo');
};
export const addDisconnectListener = (listener: Listener): void => {
  socket.on('disconnect', listener);
};

export const disconnect = (): void => {
  socket.disconnect();
};
