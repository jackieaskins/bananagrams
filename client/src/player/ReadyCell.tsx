import { CheckOutlined, CloseOutlined } from '@ant-design/icons';
import { Checkbox } from 'antd';

import { getUserId, setReady } from '../socket';
import { Player } from './types';

type ReadyCellProps = {
  player: Player;
};

const ReadyCell = ({
  player: { userId, isReady },
}: ReadyCellProps): JSX.Element => {
  const isCurrentUser = userId === getUserId();

  if (isCurrentUser) {
    return (
      <Checkbox
        checked={isReady}
        onChange={({ target: { checked } }) => {
          setReady({ isReady: checked });
        }}
      />
    );
  }

  if (isReady) {
    return <CheckOutlined css={{ color: 'green' }} />;
  }

  return <CloseOutlined css={{ color: 'red' }} />;
};
export default ReadyCell;
