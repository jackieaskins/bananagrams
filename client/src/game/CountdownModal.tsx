import { Modal } from 'antd';
import { useMemo } from 'react';

import { getUserId } from '../socket';
import { useGameCountdown, useGamePlayers, useGameStatus } from './stateHooks';

const CountdownModal = (): JSX.Element => {
  const userId = getUserId();
  const gameStatus = useGameStatus();
  const countdown = useGameCountdown();
  const players = useGamePlayers();

  const topBanana = useMemo(() => {
    const winner = players.find(({ isTopBanana }) => isTopBanana);

    if (!winner) {
      return 'Someone';
    }

    return userId === winner.userId ? 'You' : winner.username;
  }, [players, userId]);

  const modalTitle = useMemo(
    () => (gameStatus === 'STARTING' ? 'Game starting' : 'Game over'),
    [gameStatus]
  );

  const modalText = useMemo(() => {
    const countdownText = `${countdown} second${countdown === 1 ? '' : 's'}`;

    if (gameStatus === 'STARTING') {
      return `Everyone is ready, the game will start in ${countdownText}`;
    }

    return `${topBanana} won! The game will end in ${countdownText}`;
  }, [gameStatus, countdown, topBanana]);

  return (
    <Modal
      closable={false}
      footer={null}
      title={modalTitle}
      visible={['STARTING', 'ENDING'].includes(gameStatus)}
    >
      {modalText}
    </Modal>
  );
};

export default CountdownModal;
