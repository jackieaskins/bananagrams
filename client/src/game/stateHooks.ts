import { useRecoilCallback, useRecoilValue } from 'recoil';

import { initializeState } from '../game/state';
import { GameInfo } from '../games/types';
import { Player } from '../players/types';

const {
  inProgressState,
  nameState,
  playersState,
  currentPlayerState,
} = initializeState();

export const useIsGameInProgress = (): boolean =>
  useRecoilValue(inProgressState);

export const useGameName = (): string => useRecoilValue(nameState);

export const useGamePlayers = (): Player[] => useRecoilValue(playersState);

export const useCurrentPlayer = (): Player | null =>
  useRecoilValue(currentPlayerState);

export const useUpdateGameState = (): ((gameInfo: GameInfo) => void) =>
  useRecoilCallback(
    ({ set }) => (gameInfo) => {
      set(inProgressState, gameInfo.isInProgress);
      set(nameState, gameInfo.gameName);
      set(playersState, gameInfo.players);
    },
    []
  );
