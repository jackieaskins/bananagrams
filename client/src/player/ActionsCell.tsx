import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { Player } from '../players/types';
import { kickPlayer } from '../socket';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
