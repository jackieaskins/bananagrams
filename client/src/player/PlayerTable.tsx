import { Table, Typography } from 'antd';

import { useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();

  return (
    <Table
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            {
              align: COLUMN_ALIGN,
              key: 'actions',
              render: function ActionsRender(_, player) {
                return <ActionsCell player={player} />;
              },
            },
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      size="middle"
    />
  );
};

export default PlayerTable;
