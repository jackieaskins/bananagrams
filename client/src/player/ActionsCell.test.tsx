import { Button } from 'antd';
import { shallow } from 'enzyme';

import { playerFixture } from '../fixtures/player';
import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import ActionsCell from './ActionsCell';

const mockUseCurrentPlayer = useCurrentPlayer as jest.Mock;
jest.mock('../game/stateHooks', () => ({
  useCurrentPlayer: jest.fn(),
}));

const mockKickPlayer = kickPlayer as jest.Mock;
jest.mock('../socket', () => ({
  kickPlayer: jest.fn(),
}));

describe('<ActionsCell />', () => {
  const userId = '1';
  const renderComponent = (propOverrides = {}) =>
    shallow(
      <ActionsCell player={playerFixture({ userId })} {...propOverrides} />
    );

  describe('when current player is admin but not cell player', () => {
    beforeEach(() => {
      mockUseCurrentPlayer.mockReturnValue(playerFixture({ isAdmin: true }));
    });

    it('renders action button', () => {
      expect(renderComponent()).toMatchSnapshot();
    });

    it('kicks player on button click', () => {
      renderComponent().find(Button).simulate('click');

      expect(mockKickPlayer).toHaveBeenCalledWith({ userId });
    });
  });

  it('does not render if current player is admin and cell player', () => {
    mockUseCurrentPlayer.mockReturnValue(
      playerFixture({ isAdmin: true, userId })
    );

    expect(renderComponent()).toMatchSnapshot();
  });

  it('does not render if current player is not admin', () => {
    mockUseCurrentPlayer.mockReturnValue(playerFixture({ isAdmin: false }));

    expect(renderComponent()).toMatchSnapshot();
  });

  it('does not render if no current player', () => {
    mockUseCurrentPlayer.mockReturnValue(null);

    expect(renderComponent()).toMatchSnapshot();
  });
});
