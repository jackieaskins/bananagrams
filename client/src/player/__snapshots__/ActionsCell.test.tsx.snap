// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionsCell /> does not render if current player is admin and cell player 1`] = `""`;

exports[`<ActionsCell /> does not render if current player is not admin 1`] = `""`;

exports[`<ActionsCell /> does not render if no current player 1`] = `""`;

exports[`<ActionsCell /> when current player is admin but not cell player renders action button 1`] = `
<Button
  icon={<ForwardRef(DeleteOutlined) />}
  onClick={[Function]}
  type="text"
/>
`;
